lib/trino-filesystem/src/main/java/io/trino/filesystem/memory/MemoryFileSystemCache.java

```diff
@@ -0,0 +1,219 @@
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateInput");
+            return withTracing(span, delegate::newInput);
+        }
+
+        return new MemoryInput(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public TrinoInputStream cacheStream(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateStream");
+            return withTracing(span, delegate::newStream);
+        }
+
+        return new MemoryInputStream(delegate.location(), cachedEntry.get());
+    }
+
+    @Override
+    public long cacheLength(TrinoInputFile delegate, String key)
+            throws IOException
+    {
+        Optional<Slice> cachedEntry = getFromCache(key, delegate);
+        if (cachedEntry.isEmpty()) {
+            largeFileSkippedCount.incrementAndGet();
+            Span span = createSpan(key, delegate.location(), "delegateLength");
+            return withTracing(span, delegate::length);
+        }
+
+        return cachedEntry.get().length();
+    }
+
+    @Override
+    public void expire(Location location)
+            throws IOException
+    {
+        List<String> expired = cache.asMap().keySet().stream()
+                .filter(key -> key.contains(location.path()))
+                .collect(toImmutableList());
+        cache.invalidateAll(expired);
+    }
+
+    @Managed
+    public void flushCache()
+    {
+        cache.invalidateAll();
+    }
+
+    @Managed
+    public long getHitCount()
+    {
+        return cache.stats().hitCount();
+    }
+
+    @Managed
+    public long getRequestCount()
+    {
+        return cache.stats().requestCount();
+    }
+
+    @Managed
+    public long getLargeFileSkippedCount()
+    {
+        return largeFileSkippedCount.get();
+    }
+
+    @VisibleForTesting
+    boolean isCached(String key)
+    {
+        Optional<Slice> cachedEntry = cache.getIfPresent(key);
+        return cachedEntry != null && cachedEntry.isPresent();
+    }
+
+    private Optional<Slice> getFromCache(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        try {
+            // File length is available for free at this point, we add it to cache key to be resilient to manual modifications to file
+            return cache.get(key, () -> load(key, delegate));
+        }
+        catch (ExecutionException e) {
+            throw handleException(delegate.location(), e.getCause());
+        }
+    }
+
+    private Optional<Slice> load(String key, TrinoInputFile delegate)
+            throws IOException
+    {
+        Span span = createSpan(key, delegate.location(), "loadCache");
+        return withTracing(span, () -> {
+            long fileSize = delegate.length();
```

Comment:
Sometimes the caller knows the file size. Will this line do an avoidable HEAD request to S3?

Reply:
All implementations of `TrinoInputFile` cache file length. So if file size is already known, then this call is free. In iceberg case, only SNAPSHOT file doesn't have file length pre-computed.

Reply:
> All implementations of TrinoInputFile cache file length. So if file size is already known, then this call is free. 

Do the callers actually keep `TrinoInputFile` for long term? Does `TrinoInputFile.length` cache matter in practice?

Reply:
The `TrinoInputFile` is not re-used across queries. If we don't cache `TrinoInputFile.length` we will sequentially make a HEAD request to S3 for each Iceberg table used in a query.

Reply:
So with `MemoryFileSystemCache` and cache-trashing we could end-up doing extra S3 calls here. Maybe not a big issue though

Reply:
Length is pre-populated from iceberg table metadata for all metadata files other than SNAPSHOT, so this is free for most cases. For SNAPSHOT file we don't know length upfront, but the library would've asked for length regardless of cache behaviour and the length is cached in TrinoInputFile for subsequent requests. So I don't think it adds overhead.